src/core/server/plugins/discovery/plugin_discovery.test.ts

```diff
@@ -0,0 +1,483 @@
+import { logger } from '../../logging/__mocks__';
+import { PluginManifest } from './plugin_manifest_parser';
+import { discover } from './plugins_discovery';
+
+/**
+ * Resolves absolute path and escapes backslashes (used on windows systems).
+ * @param pathSegments Test path segments.
+ */
+function resolveForSnapshot(...pathSegments: string[]) {
+  return resolve(...pathSegments).replace(/\\/g, '\\\\');
+}
+
+const TEST_PATHS = {
+  scanDirs: {
+    nonEmpty: resolve('scan', 'non-empty'),
+    nonEmpty2: resolve('scan', 'non-empty-2'),
+    nonExistent: resolve('scan', 'non-existent'),
+    empty: resolve('scan', 'empty'),
+  },
+  paths: {
+    existentDir: resolve('path', 'existent-dir'),
+    existentDir2: resolve('path', 'existent-dir-2'),
+    nonDir: resolve('path', 'non-dir'),
+    nonExistent: resolve('path', 'non-existent'),
+  },
+};
+
+beforeEach(() => {
+  mockReaddir.mockImplementation((path, cb) => {
+    if (path === TEST_PATHS.scanDirs.nonEmpty) {
+      cb(null, ['1', '2-no-manifest', '3', '4-incomplete-manifest']);
+    } else if (path === TEST_PATHS.scanDirs.nonEmpty2) {
+      cb(null, ['5-invalid-manifest', '6', '7-non-dir', '8-incompatible-manifest']);
+    } else if (path === TEST_PATHS.scanDirs.nonExistent) {
+      cb(new Error('ENOENT'));
+    } else {
+      cb(null, []);
+    }
+  });
+
+  mockStat.mockImplementation((path, cb) => {
+    if (path.includes('non-existent')) {
+      cb(new Error('ENOENT'));
+    } else {
+      cb(null, { isDirectory: () => !path.includes('non-dir') });
+    }
+  });
+
+  mockReadFile.mockImplementation((path, cb) => {
+    if (path.includes('no-manifest')) {
+      cb(new Error('ENOENT'));
+    } else if (path.includes('invalid-manifest')) {
+      cb(null, Buffer.from('not-json'));
+    } else if (path.includes('incomplete-manifest')) {
+      cb(null, Buffer.from(JSON.stringify({ version: '1' })));
+    } else if (path.includes('incompatible-manifest')) {
+      cb(null, Buffer.from(JSON.stringify({ id: 'plugin', version: '1' })));
+    } else {
+      cb(null, Buffer.from(JSON.stringify({ id: 'plugin', version: '1', kibanaVersion: '1.2.3' })));
+    }
+  });
+});
+
+afterEach(() => {
+  jest.clearAllMocks();
+});
+
+test('properly scans folders and paths', async () => {
+  const { plugin$, error$ } = discover(
+    {
+      initialize: true,
+      scanDirs: Object.values(TEST_PATHS.scanDirs),
+      paths: Object.values(TEST_PATHS.paths),
+    },
+    {
+      branch: 'master',
+      buildNum: 1,
+      buildSha: '',
+      version: '1.2.3',
+    },
+    logger.get()
+  );
+
+  await expect(plugin$.pipe(toArray()).toPromise()).resolves.toMatchInlineSnapshot(`
+Array [
+  Object {
+    "manifest": Object {
+      "id": "plugin",
+      "kibanaVersion": "1.2.3",
+      "optionalPlugins": Array [],
+      "requiredPlugins": Array [],
+      "ui": false,
+      "version": "1",
+    },
+    "path": "${resolveForSnapshot(TEST_PATHS.scanDirs.nonEmpty, '1')}",
```

Comment:
__note:__ I'm not super happy about this solution, but it works on windows too and we probably aren't going to change these tests a lot so :man_shrugging:

Reply:
> Do you know what jest does here if you use the -u flag?

I'm pretty sure, jest will just override without respecting my parameters.

> I think a basic .toEqual() would be better

Okay, I think you're right, will change it to something like this:
```ts
await expect(plugin$.pipe(toArray()).toPromise()).resolves.toEqual(
    [
      resolveForSnapshot(TEST_PATHS.scanDirs.nonEmpty, '1'),
      resolveForSnapshot(TEST_PATHS.scanDirs.nonEmpty, '3'),
      resolveForSnapshot(TEST_PATHS.scanDirs.nonEmpty2, '6'),
      resolveForSnapshot(TEST_PATHS.paths.existentDir),
      resolveForSnapshot(TEST_PATHS.paths.existentDir2),
    ].map(path => ({
      manifest: {
        id: 'plugin',
        kibanaVersion: '1.2.3',
        optionalPlugins: [],
        requiredPlugins: [],
        ui: false,
        version: '1',
      },
      path,
    }))
  );
```